opentelemetry/proto/profiles/v1development/eventprofile.md

```diff
@@ -0,0 +1,284 @@
+
+The goals for the design are to be a strict superset of pprof and a useful subset of JFR, Go Execution Traces and Linux perf. Efficiency is also a goal, but it is secondary to the goal of providing a flexible and extensible format that can be implemented by a wide range of profilers and analysis tools.
+
+TODO: Finish describing the design from a high level.
+
+## Examples
+
+TODO: Add example for Go mutex and goroutine profiles.
+
+### Aggregated CPU Profile
+
+Below is a simple example of an aggregated CPU profile that demonstrates how the new format can operate in the same way as pprof.
+
+```txtpb
+event_types: [
+    { # index 0
+        name_string_index: 1 # cpu_sample
+        unit_string_index: 2 # ms
+        default_value:     10 # each Event.count represents 10ms of CPU time - this avoid redundant 
+    }
+]
+strings: [
+    "",           # index 0
+    "cpu_sample", # index 1
+    "ms"          # index 2
+]
+
+events: [
+    {
+        event_type_index: 0 # cpu_sample (can be omitted on the wire)
+        stack_index:      1 # main;serve_request
+        count:            2 # samples, i.e. 20ms of CPU time
+    }
+    {
+        event_type_index: 0 # cpu_sample (can be omitted on the wire)
+        stack_index:      2 # main;foo
+        count:            1 # samples, i.e. 10ms of CPU time
+    },
+]
+
+# stacks, locations, functions, mappings, etc. omitted for brevity
+```
+
+### Timestamped CPU Profile
+
+```txtpb
+event_types: [
+    { # index 0
+        name_string_index: 1  # cpu_sample
+        unit_string_index: 2  # ms
+        default_value:     10 # each event represents 10ms of CPU time
+        clock_id:          0
+    }
+]
+clocks: [
+    { # index 0
+        frequency :      1_000_000_000          # 1 Ghz clock, i.e. 1 cycle = 1 ns
+        time_unix_nanos: 1_257_894_000_000_000_000
+    }
+]
+attributes: [
+    { # index 0
+        key_index: 4          # thread.id
+        value: {int_value: 1} # 1
+    },
+    { # index 1
+        key_index: 4,         # thread.id
+        value: {int_value: 2} # 2
+    }
+]
+strings: [
+    "",             # index 0
+    "cpu_time",     # index 1
+    "milliseconds", # index 2
+    "unix_epoch",   # index 3
+    "thread.id"     # index 4
+]
+
+events: [
+    {
+        event_type_index: 0     # cpu_sample (can be omitted on the wire)
+        stack_index: 1          # main;serve_request
+        time:        10_500_000 # 10.5ms since clock start
+        attributes:  [0]        # thread_id: 1
+    }
+    {
+        event_type_index: 0  # cpu_sample (can be omitted on the wire)
+        stack_index: 2       # main;foo
+        time:        300_000 # 0.3ms since previous event
```

Comment:
Why is time based on previous event? This implies we have events in perfect time order. Often during collection, especially off-CPU (cswitch) events, the start of the event is not in-order of when it completes. This would cause collectors to sort potentially large sets of data (cswitches can be massive).

Can we just have an absolute time option vs the delta time encoding scheme?

Reply:
I imagine delta encoding is used to reduce payload size

Reply:
The purpose of delta encoding is to reduce payload sizes. `time` is an `int64` right now, so negative values are possible. If we frequently need to deal with unordered events, we could define `time` as an `sint64` which allows to encode negative numbers more efficiently. However, it comes a the expense of encoding positive numbers.

Anyway, I don't feel very strongly about delta encoding. I'm planning to do some benchmarks later to provide some evidence (or lack of evidence) that it's going to be beneficial.

Reply:
64-bits are already variable encoded, and if they were deltas off of a stable epoch (like profile start time). They would be quite small without deltas between each other. I get why we would want delta encoding, but it comes at a sorting cost that is likely really bad for off-cpu data sources (since they complete at random times compared to the start time).